### PR TITLE
Updates faulty MG 'status' and provides error messages

### PR DESCRIPTION
**Pull request by aarya-bhatia**
- Added a 'status' and 'message' key to every MG. 
- The status is set to 0 by default. On any kind of error, it changes to 1. 
- The message key contains the description of the error message. 
- There are similar changes in the server_manager file which remove faulty MG from the 'names' and 'weights' list inside the update function, (if the update data contains a non-zero status). 
- [EDIT] If the generate segment returns a 500 response (and updates the MG status), then we retry until a valid MG is found. Only if there are none, we send a 500 response to the front-end.
- I am not storing HTTP status codes but we could store them in the DB too if required. 
- I am handling up to 4 kinds of errors: Timeout errors, TooManyRedirects error, a generic RequestException, and my custom error, if the maze has a invalid format, (maze.width % 7 != 0 or maze.height % 7 != 0). Furthermore, I try to decode the given maze data using my maze package, if there is an error it will throw an exception (Eg. Some row has a missing column, etc). Even this kind of error is handled. 
- There are possibilities for further error handling in the /generateSegment route...
- The kill-switch file has a script to remove all Mgs from production that has an incorrect URL (containing 127.0.0.1).
- small changes to my personal MGs.

### CODE REVIEW

**Comment by pcrain** [comment]
I am trying to run this code locally and getting the following error:

```
...
    if doc['status'] == 0:
KeyError: 'status'
...
```

Any ideas what could be going wrong here or how to resolve it? :open_mouth:

**Comment by aarya-bhatia** [comment]
@pcrain I think I do know why you got that error. I added the new status key to all MGs but you loaded existing MGs from the database which did not have that key yet. I am fixing this issue by adding a status key if there is not one already.

**Comment by PeterYR** [comment]
There doesn't seem to be a way to re-add an MG that failed. When I kill my MG, the middleware marks it as faulty (with `Error: Request Exception` on the server list page) and gives a 503 as expected, but when I re-send the request to add the MG, the middleware doesn't seem to update its list of available MGs, and still gives a 503 to the front-end.

The middleware sends a `400` when I try to re-add my MG, so I suspect it's got something to do with line 130 in `servers.py`.

File: .env [hunk 1]
```diff
@@ -1 +1 @@
-FLASK_ENV=production
\ No newline at end of file
+FLASK_ENV=development
```
Comment (PeterYR) [inline .env:1]: It'd probably be best to leave this on `production` mode. `development` mode seems to impact performance.

**Comment by aarya-bhatia** [comment]
> There doesn't seem to be a way to re-add an MG that failed. When I kill my MG, the middleware marks it as faulty (with `Error: Request Exception` on the server list page) and gives a 503 as expected, but when I re-send the request to add the MG, the middleware doesn't seem to update its list of available MGs, and still gives a 503 to the front-end.
> 
> The middleware sends a `400` when I try to re-add my MG, so I suspect it's got something to do with line 130 in `servers.py`.

Are you trying to re-add your MG with the /addMG route? If so, I think that is the issue. You need to call the /updateMG route as addMG will not let you add duplicate name MGs.

File: .env (hunk 1 above)
Comment (aarya-bhatia) [reply .env:1]: You're right I did not intend to push that change, I prefer the development while running flask locally as it reloads on save.

**Comment by PeterYR** [comment]
> > There doesn't seem to be a way to re-add an MG that failed. When I kill my MG, the middleware marks it as faulty (with `Error: Request Exception` on the server list page) and gives a 503 as expected, but when I re-send the request to add the MG, the middleware doesn't seem to update its list of available MGs, and still gives a 503 to the front-end.
> > The middleware sends a `400` when I try to re-add my MG, so I suspect it's got something to do with line 130 in `servers.py`.
> 
> Are you trying to re-add your MG with the /addMG route? If so, I think that is the issue. You need to call the /updateMG route as addMG will not let you add duplicate name MGs.

Ah, makes sense. What about making the `/addMG` behave like `/updateMG` if the MG has already been added? It'd make things simpler for MGs that add themselves on startup.

**Comment by aarya-bhatia** [comment]
> > > There doesn't seem to be a way to re-add an MG that failed. When I kill my MG, the middleware marks it as faulty (with `Error: Request Exception` on the server list page) and gives a 503 as expected, but when I re-send the request to add the MG, the middleware doesn't seem to update its list of available MGs, and still gives a 503 to the front-end.
> > > The middleware sends a `400` when I try to re-add my MG, so I suspect it's got something to do with line 130 in `servers.py`.
> > 
> > 
> > Are you trying to re-add your MG with the /addMG route? If so, I think that is the issue. You need to call the /updateMG route as addMG will not let you add duplicate name MGs.
> 
> Ah, makes sense. What about making the `/addMG` behave like `/updateMG` if the MG has already been added? It'd make things simpler for MGs that add themselves on startup.

Sure, I can do that

**Comment by pcrain** [comment]
Looks good so far. :smile: We'd ideally like to merge this code into the main code base by tomorrow morning, so please let me know whenever you are finished. Thanks! :smiley:

**Comment by aarya-bhatia** [comment]
@pcrain  I think I am done. In my last commit, I send one of the default MGs if none are available, instead of a 503 response, to the frontend.

**Comment by aarya-bhatia** [comment]
I am trying to merge some conflicts

**Comment by pcrain** [comment]
Looks good to me, thank you both for all of your hard work! :smiley:

**Pull request merged**